### PR TITLE
En 9856 waiting epochs fix

### PR DESCRIPTION
**Pull request by raduchis**
1. fixed problem with some validators having 6 epochs in the waiting lists because of other leaving nodes
2. added allow to remove more than 80 nodes from the waiting lists

For testings:
for 1. unstake 1 node from eligible and 1 node from waiting and now the entire allowed number of validators that can be shuffled out should enter the eligible (on the mainnet that number is 80)
for 2. unstake more than the maxNodesToShuffle (80 nodes on the mainnet) that are from the waiting list of a shard and it should work

### CODE REVIEW

File: sharding/hashValidatorShuffler.go [hunk 1]
```diff
@@ -354,13 +372,33 @@ func removeLeavingNodesFromValidatorMaps(
...
 
 	stillRemainingInLeaving := make([]Validator, len(leaving))
 	copy(stillRemainingInLeaving, leaving)
 
-	waiting, stillRemainingInLeaving = removeNodesFromMap(waiting, stillRemainingInLeaving, numToRemove)
-	eligible, stillRemainingInLeaving = removeNodesFromMap(eligible, stillRemainingInLeaving, numToRemove)
+	if !waitingFixEnabled {
+		waiting, stillRemainingInLeaving = removeNodesFromMap(waiting, stillRemainingInLeaving, numToRemove)
+		eligible, stillRemainingInLeaving = removeNodesFromMap(eligible, stillRemainingInLeaving, numToRemove)
+	} else {
```
Comment (iulianpascalau) [inline sharding/hashValidatorShuffler.go:385]: early return here, remove else branch

File: sharding/hashValidatorShuffler.go [hunk 2]
```diff
@@ -354,13 +372,33 @@ func removeLeavingNodesFromValidatorMaps(
...
+		maxNumToRemoveFromWaiting := make(map[uint32]int)
+		for i := range numToRemove {
+			maxNumToRemoveFromWaiting[i] = len(eligible[i]) + len(waiting[i]) - minNodesPerShard
```
Comment (iulianpascalau) [inline sharding/hashValidatorShuffler.go:388]: as discussed, maybe add a check here to not get into negative values.

File: sharding/hashValidatorShuffler.go (hunk 1 above)
Comment (raduchis) [reply sharding/hashValidatorShuffler.go:385]: fixed

File: sharding/hashValidatorShuffler.go (hunk 2 above)
Comment (raduchis) [reply sharding/hashValidatorShuffler.go]: added check

File: sharding/hashValidatorShuffler.go [hunk 3]
```diff
@@ -380,26 +383,54 @@ func removeLeavingNodesFromValidatorMaps(
...
+	eligible map[uint32][]Validator,
+	waiting map[uint32][]Validator,
+	numToRemove map[uint32]int,
+	stillRemainingInLeaving []Validator,
+	minNodesMeta int,
+	minNodesPerShard int,
+) (map[uint32][]Validator, map[uint32][]Validator, []Validator) {
+	maxNumToRemoveFromWaiting := make(map[uint32]int)
+	for shardId := range numToRemove {
+		minimumNumberOfNodes := minNodesPerShard
```
Comment (iulianpascalau) [inline sharding/hashValidatorShuffler.go:404]: Duplicated code with L418. 
Maybe we can figure out a way to refactor these components as to remove the number of passed parameters and the duplicated code.

**Review by iulianpascalau — APPROVED** [review]

File: sharding/hashValidatorShuffler.go [hunk 4]
```diff
@@ -354,13 +372,33 @@ func removeLeavingNodesFromValidatorMaps(
...
 	copy(stillRemainingInLeaving, leaving)
 
-	waiting, stillRemainingInLeaving = removeNodesFromMap(waiting, stillRemainingInLeaving, numToRemove)
-	eligible, stillRemainingInLeaving = removeNodesFromMap(eligible, stillRemainingInLeaving, numToRemove)
+	if !waitingFixEnabled {
+		waiting, stillRemainingInLeaving = removeNodesFromMap(waiting, stillRemainingInLeaving, numToRemove)
+		eligible, stillRemainingInLeaving = removeNodesFromMap(eligible, stillRemainingInLeaving, numToRemove)
+	} else {
+		maxNumToRemoveFromWaiting := make(map[uint32]int)
+		for i := range numToRemove {
```
[the remark this answers is not in the document]
Comment (raduchis) [reply sharding/hashValidatorShuffler.go]: fixed, iterated over eligible

**Review by AdoAdoAdo — DISMISSED** [review]

File: sharding/hashValidatorShuffler.go [hunk 5]
```diff
@@ -430,7 +430,7 @@ func removeLeavingNodes(
 	}
 
 	newEligible, stillRemainingInLeaving := removeNodesFromMap(eligible, stillRemainingInLeaving, numToRemove)
-	return newWaiting, newEligible, stillRemainingInLeaving
+	return newEligible, newWaiting, stillRemainingInLeaving
```
Comment (iulianpascalau) [inline sharding/hashValidatorShuffler.go:433]: returning the same type of variables problem

**Review by iulianpascalau — APPROVED** [review]

File: sharding/hashValidatorShuffler.go (hunk 3 above)
Comment (raduchis) [reply sharding/hashValidatorShuffler.go:404]: fixed

File: sharding/indexHashedNodesCoordinator.go [hunk 6]
```diff
@@ -662,8 +692,23 @@ func (ihgs *indexHashedNodesCoordinator) computeNodesConfigFromList(
 			eligibleMap[validatorInfo.ShardId] = append(eligibleMap[validatorInfo.ShardId], currentValidator)
 		case string(core.LeavingList):
 			log.Debug("leaving node trie", "pk", validatorInfo.PublicKey)
-			eligibleMap[validatorInfo.ShardId] = append(eligibleMap[validatorInfo.ShardId], currentValidator)
 			leavingMap[validatorInfo.ShardId] = append(leavingMap[validatorInfo.ShardId], currentValidator)
+			if !ihgs.flagWaitingListFix.IsSet() {
```
Comment (iulianpascalau) [inline sharding/indexHashedNodesCoordinator.go:696]: please extract methods here

File: sharding/indexHashedNodesCoordinator_test.go [hunk 7]
```diff
@@ -1728,6 +1753,38 @@ func TestIndexHashedNodesCoordinator_ShuffleOut_NilConfig(t *testing.T) {
 	require.Equal(t, expectedShardForNotFound, newShard)
 }
 
+func TestIndexHashedNodesCoordinator_computeNodesConfigFromList_NilPreviousNodesConfig(t *testing.T) {
```
Comment (iulianpascalau) [inline sharding/indexHashedNodesCoordinator_test.go:1756]: Please remove the second _

File: sharding/indexHashedNodesCoordinator_test.go [hunk 8]
```diff
@@ -1774,32 +1831,35 @@ func TestIndexHashedNodesCoordinator_computeNodesConfigFromList_NilPk(t *testing
...
-	newNodesConfig, err := ihgs.computeNodesConfigFromList(validatorInfos)
+	newNodesConfig, err := ihgs.computeNodesConfigFromList(&epochNodesConfig{}, validatorInfos)
 
 	assert.Nil(t, newNodesConfig)
 	assert.NotNil(t, err)
 	assert.Equal(t, ErrNilPubKey, err)
 }
 
-func TestIndexHashedNodesCoordinator_computeNodesConfigFromList_Validators(t *testing.T) {
+func TestIndexHashedNodesCoordinator_computeNodesConfigFromList_ValidatorsWithFix(t *testing.T) {
```
Comment (iulianpascalau) [inline sharding/indexHashedNodesCoordinator_test.go:1841]: remove the second _

File: sharding/indexHashedNodesCoordinator_test.go [hunk 9]
```diff
@@ -1842,25 +1906,158 @@ func TestIndexHashedNodesCoordinator_computeNodesConfigFromList_Validators(t *te
...
+	leavingListMeta := createValidatorList(ihgs,
+		[]*state.ShardValidatorInfo{shardMetaLeaving1})
+	assert.Equal(t, leavingListMeta, newNodesConfig.leavingMap[core.MetachainShardId])
+
+	newListShardZero := createValidatorList(ihgs,
+		[]*state.ShardValidatorInfo{shard0New0})
+	assert.Equal(t, newListShardZero, newNodesConfig.newList)
+}
+
+func TestIndexHashedNodesCoordinator_computeNodesConfigFromList_ValidatorsNoFix(t *testing.T) {
```
Comment (iulianpascalau) [inline sharding/indexHashedNodesCoordinator_test.go:1970]: remove the second _

File: sharding/indexHashedNodesCoordinator.go [hunk 10]
```diff
@@ -662,8 +692,23 @@ func (ihgs *indexHashedNodesCoordinator) computeNodesConfigFromList(
...
+				eligibleMap[validatorInfo.ShardId] = append(eligibleMap[validatorInfo.ShardId], currentValidator)
+			} else {
+				found, shardId := searchInMap(previousEpochConfig.eligibleMap, currentValidator.PubKey())
+				if found {
+					log.Debug("leaving node trie found in", "list", "eligible", "shardId", shardId)
+					eligibleMap[shardId] = append(eligibleMap[validatorInfo.ShardId], currentValidator)
```
Comment (iulianpascalau) [inline sharding/indexHashedNodesCoordinator.go:702]: Is this correct? Why do we re-add a leaving validator to the eligible/waiting map if it was included in a previous epoch configuration as an eligible/waiting ?

**Review by iulianpascalau — CHANGES_REQUESTED** [review]

File: epochStart/bootstrap/disabled/disabledEpochNotifier.go [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+}
+
+// NewEpochNotifier returns the disabled epoch start notifier
+func NewEpochNotifier() *EpochNotifier {
+	return &EpochNotifier{}
+}
+
+// CheckEpoch -
+func (en *EpochNotifier) CheckEpoch(_ uint32) {
+
```
[the remark this answers is not in the document]
Comment (raduchis) [reply epochStart/bootstrap/disabled/disabledEpochNotifier.go:18]: fixed

File: epochStart/bootstrap/syncValidatorStatus.go [hunk 12]
```diff
@@ -98,6 +98,7 @@ func NewSyncValidatorStatus(args ArgsNewSyncValidatorStatus) (*syncValidatorStat
 		SelfPublicKey:           args.PubKey,
 		ConsensusGroupCache:     consensusGroupCache,
 		ShuffledOutHandler:      disabled.NewShuffledOutHandler(),
+		EpochNotifier:           disabled.NewEpochNotifier(),
```
[the remark this answers is not in the document]
Comment (raduchis) [reply epochStart/bootstrap/syncValidatorStatus.go:101]: good catch! fixed

File: cmd/storer2elastic/dataprocessor/dataProcessor.go [hunk 13]
```diff
@@ -279,6 +279,7 @@ func (dp *dataProcessor) createNodesCoordinatorForShard(nodesConfig update.Genes
 		SelfPublicKey:           []byte("own public key"),
 		ConsensusGroupCache:     consensusGroupCache,
 		ShuffledOutHandler:      disabled.NewShuffledOutHandler(),
+		EpochNotifier:           disabled.NewEpochNotifier(),
```
[the remark this answers is not in the document]
Comment (raduchis) [reply cmd/storer2elastic/dataprocessor/dataProcessor.go:282]: fixed

File: sharding/indexHashedNodesCoordinator.go [hunk 14]
```diff
@@ -125,6 +129,8 @@ func NewIndexHashedNodesCoordinator(arguments ArgNodesCoordinator) (*indexHashed
 		shuffledOutHandler:            arguments.ShuffledOutHandler,
 		startEpoch:                    arguments.StartEpoch,
 		publicKeyToValidatorMap:       make(map[string]*validatorWithShardID),
+		waitingListFixEnableEpoch:     arguments.WaitingListFixEnabledEpoch,
+		epochNotifier:                 arguments.EpochNotifier,
```
[the remark this answers is not in the document]
Comment (raduchis) [reply sharding/indexHashedNodesCoordinator.go:133]: removed

File: sharding/indexHashedNodesCoordinator.go [hunk 15]
```diff
@@ -1096,3 +1141,8 @@ func createValidatorInfoFromBody(
 
 	return allValidatorInfo, nil
 }
+
+func (ihgs *indexHashedNodesCoordinator) EpochConfirmed(epoch uint32) {
```
[the remark this answers is not in the document]
Comment (raduchis) [reply sharding/indexHashedNodesCoordinator.go:1145]: fixed

File: sharding/indexHashedNodesCoordinator.go (hunk 6 above)
Comment (raduchis) [reply sharding/indexHashedNodesCoordinator.go:696]: fixed

File: sharding/indexHashedNodesCoordinator_test.go (hunk 9 above)
Comment (raduchis) [reply sharding/indexHashedNodesCoordinator_test.go:1970]: fixed

File: sharding/indexHashedNodesCoordinator_test.go (hunk 8 above)
Comment (raduchis) [reply sharding/indexHashedNodesCoordinator_test.go:1841]: fixed

File: sharding/indexHashedNodesCoordinator_test.go (hunk 7 above)
Comment (raduchis) [reply sharding/indexHashedNodesCoordinator_test.go:1756]: fixed

File: sharding/shardingArgs.go [hunk 16]
```diff
@@ -8,20 +8,22 @@ import (
...
+	ShardIDAsObserver          uint32
+	NbShards                   uint32
+	EligibleNodes              map[uint32][]Validator
+	WaitingNodes               map[uint32][]Validator
+	SelfPublicKey              []byte
+	Epoch                      uint32
+	StartEpoch                 uint32
+	ConsensusGroupCache        Cacher
+	ShuffledOutHandler         ShuffledOutHandler
+	EpochNotifier              EpochNotifier
```
Comment (AdoAdoAdo) [inline sharding/shardingArgs.go:27]: epoch notifier not needed by nodes coordinator

File: sharding/indexHashedNodesCoordinatorWithRater_test.go [hunk 17]
```diff
@@ -805,6 +812,7 @@ func BenchmarkIndexHashedWithRaterGroupSelector_ComputeValidatorsGroup21of400(b
 		SelfPublicKey:           []byte("key"),
 		ConsensusGroupCache:     &mock.NodesCoordinatorCacheMock{},
 		ShuffledOutHandler:      &mock.ShuffledOutHandlerStub{},
+		EpochNotifier:           &mock.EpochNotifierStub{},
```
Comment (AdoAdoAdo) [inline sharding/indexHashedNodesCoordinatorWithRater_test.go:815]: epoch notifier changes can be reverted

File: sharding/indexHashedNodesCoordinator.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (AdoAdoAdo) [reply sharding/indexHashedNodesCoordinator.go:1145]: rename this e.g UpdateEpochFlags(epoch uint32) and call it on the EpochStartPrepare with the metablock epoch.

File: epochStart/bootstrap/syncValidatorStatus.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (raduchis) [reply epochStart/bootstrap/syncValidatorStatus.go:101]: removed actually

File: sharding/indexHashedNodesCoordinator.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (raduchis) [reply sharding/indexHashedNodesCoordinator.go:1145]: fixed

File: sharding/indexHashedNodesCoordinator.go [hunk 18]
```diff
@@ -662,8 +692,23 @@ func (ihgs *indexHashedNodesCoordinator) computeNodesConfigFromList(
...
 		case string(core.LeavingList):
 			log.Debug("leaving node trie", "pk", validatorInfo.PublicKey)
-			eligibleMap[validatorInfo.ShardId] = append(eligibleMap[validatorInfo.ShardId], currentValidator)
 			leavingMap[validatorInfo.ShardId] = append(leavingMap[validatorInfo.ShardId], currentValidator)
+			if !ihgs.flagWaitingListFix.IsSet() {
+				eligibleMap[validatorInfo.ShardId] = append(eligibleMap[validatorInfo.ShardId], currentValidator)
+			} else {
+				found, shardId := searchInMap(previousEpochConfig.eligibleMap, currentValidator.PubKey())
+				if found {
+					log.Debug("leaving node trie found in", "list", "eligible", "shardId", shardId)
```
[the remark this answers is not in the document]
Comment (raduchis) [reply sharding/indexHashedNodesCoordinator.go:701]: fixed

File: sharding/indexHashedNodesCoordinator.go [hunk 19]
```diff
@@ -85,6 +86,9 @@ type indexHashedNodesCoordinator struct {
 	shuffledOutHandler            ShuffledOutHandler
 	startEpoch                    uint32
 	publicKeyToValidatorMap       map[string]*validatorWithShardID
+	waitingListFixEnableEpoch     uint32
+	flagWaitingListFix            atomicFlags.Flag
+	epochNotifier                 core.EpochNotifier
```
[the remark this answers is not in the document]
Comment (raduchis) [reply sharding/indexHashedNodesCoordinator.go:91]: removed

File: sharding/interface.go [hunk 20]
```diff
@@ -159,6 +160,14 @@ type EpochStartActionHandler interface {
 	NotifyOrder() uint32
 }
 
+// EpochNotifier can notify upon an epoch change and provide the current epoch
+type EpochNotifier interface {
```
[the remark this answers is not in the document]
Comment (raduchis) [reply sharding/interface.go:164]: removed

File: sharding/mock/epochNotifierStub.go [hunk 21]
```diff
@@ -0,0 +1,45 @@
+package mock
```
[the remark this answers is not in the document]
Comment (raduchis) [reply sharding/mock/epochNotifierStub.go:1]: removed

File: sharding/shardingArgs.go (hunk 16 above)
Comment (raduchis) [reply sharding/shardingArgs.go:27]: removed

File: sharding/indexHashedNodesCoordinatorWithRater_test.go (hunk 17 above)
Comment (raduchis) [reply sharding/indexHashedNodesCoordinatorWithRater_test.go:815]: fixed

File: integrationTests/nodesCoordinatorFactory.go [hunk 22]
```diff
@@ -107,6 +108,7 @@ func (ihncrf *IndexHashedNodesCoordinatorWithRaterFactory) CreateNodesCoordinato
 		ConsensusGroupCache:     arg.consensusGroupCache,
 		BootStorer:              arg.bootStorer,
 		ShuffledOutHandler:      &mock.ShuffledOutHandlerStub{},
+		EpochNotifier:           &mock.EpochNotifierStub{},
```
[the remark this answers is not in the document]
Comment (raduchis) [reply integrationTests/nodesCoordinatorFactory.go:111]: fixed

File: integrationTests/testP2PNode.go [hunk 23]
```diff
@@ -303,6 +303,7 @@ func CreateNodesWithTestP2PNodes(
 			Epoch:                   0,
 			EpochStartNotifier:      notifier.NewEpochStartSubscriptionHandler(),
 			ShuffledOutHandler:      &mock.ShuffledOutHandlerStub{},
+			EpochNotifier:           &mock.EpochNotifierStub{},
```
[the remark this answers is not in the document]
Comment (raduchis) [reply integrationTests/testP2PNode.go:306]: fixed

[the remark this answers is not in the document]
Comment (raduchis) [reply integrationTests/testP2PNode.go:306]: fixed

File: epochStart/bootstrap/disabled/disabledEpochNotifier.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (raduchis) [reply epochStart/bootstrap/disabled/disabledEpochNotifier.go:18]: removed entire struct

File: integrationTests/nodesCoordinatorFactory.go (hunk 22 above)
[the remark this answers is not in the document]
Comment (raduchis) [reply integrationTests/nodesCoordinatorFactory.go:111]: fixed

File: sharding/indexHashedNodesCoordinator.go [hunk 24]
```diff
@@ -641,13 +662,22 @@ func (ihgs *indexHashedNodesCoordinator) GetChance(_ uint32) uint32 {
...
 func (ihgs *indexHashedNodesCoordinator) computeNodesConfigFromList(
+	previousEpochConfig *epochNodesConfig,
 	validatorInfos []*state.ShardValidatorInfo,
 ) (*epochNodesConfig, error) {
 	eligibleMap := make(map[uint32][]Validator)
 	waitingMap := make(map[uint32][]Validator)
 	leavingMap := make(map[uint32][]Validator)
 	newNodesList := make([]Validator, 0)
 
+	if ihgs.flagWaitingListFix.IsSet() && previousEpochConfig == nil {
```
[the remark this answers is not in the document]
Comment (raduchis) [reply sharding/indexHashedNodesCoordinator.go:673]: tested with activation epoch 0 and also multiple epochs with shuffle and works.

File: sharding/indexHashedNodesCoordinator.go [hunk 25]
```diff
@@ -1142,7 +1155,8 @@ func createValidatorInfoFromBody(
 	return allValidatorInfo, nil
 }
 
-func (ihgs *indexHashedNodesCoordinator) EpochConfirmed(epoch uint32) {
+// UpdateEpochFlags sets the flags based on the current epoch
+func (ihgs *indexHashedNodesCoordinator) UpdateEpochFlags(epoch uint32) {
```
Comment (iulianpascalau) [inline sharding/indexHashedNodesCoordinator.go:1159]: do we need to keep this exported?

File: sharding/indexHashedNodesCoordinator.go [hunk 26]
```diff
@@ -747,6 +734,32 @@ func (ihgs *indexHashedNodesCoordinator) computeNodesConfigFromList(
 	return newNodesConfig, nil
 }
 
+func (ihgs *indexHashedNodesCoordinator) addValidatorToPreviousMap(
+	previousEpochConfig *epochNodesConfig,
+	eligibleMap map[uint32][]Validator,
+	waitingMap map[uint32][]Validator,
+	currentValidator *validator,
+	currentValidatorShardId uint32) {
+	if !ihgs.flagWaitingListFix.IsSet() {
```
Comment (iulianpascalau) [inline sharding/indexHashedNodesCoordinator.go:743]: please add an empty line here

Comment (raduchis) [reply sharding/indexHashedNodesCoordinator.go:743]: fixed

**Review by AdoAdoAdo — APPROVED** [review]

**Review by iulianpascalau — APPROVED** [review]

**Review by iulianpascalau — APPROVED** [review]

**Review by AdoAdoAdo — APPROVED** [review]

**Review by gabi-vuls — APPROVED** [review]
System test passed

**Pull request merged**